### PR TITLE
Improve error message when a file to be downloaded is missing

### LINKED ISSUE #1003: Improve error message when a file to be downloaded is missing
If I file isn't present at the url when we try to download it the error that is reported isn't very useful:

```
cannot import name 'urlretrieve'
```

We get this in part because of joint support for Python 2/3 (we try the Python 3 call, fall back to Python 2, and so if issue isn't Python 3 vs 2 we get an unhelpful error message), but we need to catch this better and report something more useful to the user.

See, e.g., https://github.com/weecology/retriever/issues/865

### PR DESCRIPTION
**Pull request by apoorvaeternity**
Fixes #1003

### CODE REVIEW

**Comment by apoorvaeternity** [comment]
@ethanwhite The AppVeyor build is failing.

File: retriever/__main__.py [hunk 1]
```diff
@@ -174,11 +174,12 @@ def main():
                     dataset.engine.final_cleanup()
                 except KeyboardInterrupt:
                     pass
-                except Exception as e:
-                    print(e)
+                except Exception:
```
Comment (henrykironde) [inline retriever/__main__.py]: I think its better if we also (print e)

**Comment by henrykironde** [comment]
@apoorvaeternity thank you for the contribution. Sorry for the late reply.
I have fixed the issue that you reported `The AppVeyor build is failing.`in #1031. Next time you could create an issue. Feel free to review the PR. 

I  have reviewed your work and the main download function is https://github.com/apoorvaeternity/retriever/blob/8c6840654ba4319877c7399ce9de02c379ef8bb6/retriever/lib/engine.py#L407

**Comment by apoorvaeternity** [comment]
@henrykironde Okay so this means I have to make the changes in `engine.py` and not in `__main__.py`. Right?

**Comment by henrykironde** [comment]
Yes

**Comment by henrykironde** [comment]
If you debug using `PyCharm` or `pdb` you will notice that you hit that download function in engine first. In order to test or reproduce the error, I recommend you go to your script directory and change the urls to some wrong url for just one script. For example the Iris dataset,
https://github.com/weecology/retriever/blob/master/scripts/iris.json, is easy to work with.
Change the urls to wrong values and then try to install it `retriever install sqlite iris`. remember before you install the dataset, run `retriever reset all` to clean the environment

**Comment by apoorvaeternity** [comment]
@henrykironde How does it look now?

File: retriever/lib/engine.py [hunk 2]
```diff
@@ -406,19 +406,28 @@ def database_name(self, name=None):
...
             except:
                 # For some urls lacking filenames urlretrieve from the future
                 # package seems to fail. This issue occurred in the PlantTaxonomy
                 # script. If this happens, fall back to the standard Python 2 version.
                 from urllib import urlretrieve as py2urlretrieve
-                py2urlretrieve(url, path, reporthook=reporthook)
+                try:
+                    py2urlretrieve(url, path, reporthook=reporthook)
+                except:
+                    pass
```
Comment (henrykironde) [inline retriever/lib/engine.py]: Would replacing `print("Unable to download dataset {}".format(filename))`  for pass in line 425 , not be enough to catch the errors?

Comment (henrykironde) [reply retriever/lib/engine.py]: if you make changes would you also update to current master.
1. update master
`git fetch upstream`
`git merge upstream master`


2. In your current branch
`git rebase master`

Comment (apoorvaeternity) [reply retriever/lib/engine.py]: If I replace line 425 with ```print("Unable to download dataset {}".format(filename))``` it just prints ```cannot import name 'urlretrieve'``` and not the message.

**Comment by henrykironde** [comment]
Thanks @apoorvaeternity , am testing this in the next 5 hours

**Comment by apoorvaeternity** [comment]
@henrykironde Any updates on this?

**Comment by zhangcandrew** [comment]
Hey @apoorvaeternity!

So I went through and looked at the code, there might be a few things that need to be fixed. Also you might want to update your code base to the retriever repository master branch, since we just put in a huge change on how we deal with scripts (although, I think your PR is not affected much by this, it'd still be good so that you can test changes locally). 

Also, as @henrykironde mentioned, I think some of the boolean values like downloaded are unneeded since there is a flow already present in the try-except blocks. I know you mentioned that some things weren't printing, it might be because your excepts are too generic. Try excepting like this: 

```
except ImportError as e
```
Or something along those lines, since we know the problem with the urlretrieve is an importError. Just changing that locally for me let retriever print out slightly more verbose errors which is what I think #1003 was asking for.

**Comment by apoorvaeternity** [comment]
@zhangcandrew Okay. I will make the changes by today.

**Comment by apoorvaeternity** [comment]
@zhangcandrew You are right. I have made the changes. Just using `except ImportError` prints a better message `HTTP Error 404: Not Found`. I have also rebased on top of master branch. Kindly review.

**Comment by zhangcandrew** [comment]
Thanks!

That seems to do the trick, although I'm curious as to why it is caught by ImportError and not URLError as this documentation suggests it should be: https://docs.python.org/3.0/library/urllib.request.html#urllib.request.urlopen. 

Merging this in for now. @apoorvaeternity, if you find that problems arise because we are excepting importError instead of URLError or some of the other errors mentioned with urllib (HTTPError, etc), feel free to open this again, we can discuss and put in another PR!

**Pull request merged**